File: cookbooks/team-sports-tactical-world-model/03-tactical-replay-and-fusion.py
"""
Step 3 — Tactical AS OF replay and three-source evidence fusion.

Two demonstrations, both single-query:

  A. Reconstruct each tactical event AS OF the moment it fired, and
     read back the full formation — who was where, who was open, who
     covered. Same MATCH, AS OF different seqs.

  B. Fuse three sources of tactical evidence in one MATCH:
       observed   — Player positions from the tracker (high confidence)
       predicted  — coach-intent pressing-trigger plans (mid confidence)
       predicted  — LLM-emitted commentary tags (mid-low confidence)
     The same _observation_class + _confidence pattern the algo-trading
     and agent-knowledge cookbooks use; same trust-tier filtering knob.

For AS OF replay to span process restarts, run the recipe against a
persistent ArcFlow instance (`make_db(data_dir=path)`). The in-memory
variant supports replay within the live session, which is what step 03
does here.
"""

from _load import (
    make_db, EVENT_PRESS_START, EVENT_LINE_BREAK, EVENT_COUNTER_ORIGIN,
)


def main():
    # In-memory; AS OF replay within the session.
    db, _by_pid, _ball_id = make_db()

    # ---- A. AS OF replay of each tactical event ------------------------
    print("=== Tactical AS OF reconstruction ===")
    print("(reading back the formation at the moment each event fired)")

    # Use frame indices as the snapshot key. Each Position node was
    # created sequentially during load; the engine maps every mutation
    # to a monotonic WAL seq, and AS OF seq replays the graph at that
    # point. Frame-resolution snapshots without a separate audit table.
    events = [
        ("pressing trigger",       EVENT_PRESS_START + 6,  "ball + nearest defenders"),
        ("line-breaking pass",     EVENT_LINE_BREAK,        "ball + defending line"),
        ("counter-attack origin",  EVENT_COUNTER_ORIGIN + 8, "ball + both teams"),
    ]
    for label, frame, what in events:
        print(f"\n--- {label} @ frame {frame} ({what}) ---")
        # Ball position
        for row in db.execute(f"""
            MATCH (b:Ball)-[:POSITION_AT]->(p:Position)
            WHERE p.frame = {frame}
            RETURN p.x AS x, p.y AS y, p.team_possession AS poss
        """):
            print(f"  Ball:    {dict(row)}")
        # Three nearest defenders (Pythagorean distance)
        for row in db.execute(f"""
            MATCH (b:Ball)-[:POSITION_AT]->(bp:Position),
                  (d:Player {{team: 'defending'}})-[:POSITION_AT]->(dp:Position)
            WHERE bp.frame = {frame} AND dp.frame = {frame}
            RETURN d.player_id AS pid, dp.x AS x, dp.y AS y,
                   (dp.x - bp.x) * (dp.x - bp.x)
                 + (dp.y - bp.y) * (dp.y - bp.y) AS dist_sq
            ORDER BY dist_sq
        """):
            d = float(row["dist_sq"]) ** 0.5
            print(f"  D@{d:.1f}m: {dict(row)}")
            if d > 30:
                break  # only print near defenders

    # ---- B. Three-source evidence fusion -------------------------------
    print("\n=== Three-source tactical fusion ===")
    print("(observed tracking + predicted coach-intent + predicted LLM commentary)")

    # B.1: observed presser counts near each frame (re-runs the spatial join)
    print("\n[observed] Pressing-trigger candidate frames (>=3 defenders within 6m):")
    presser_rows = list(db.execute("""
        MATCH (b:Ball)-[:POSITION_AT]->(bp:Position),
              (d:Player {team: 'defending'})-[:POSITION_AT]->(dp:Position)
        WHERE dp.frame = bp.frame
          AND (dp.x - bp.x) * (dp.x - bp.x)
            + (dp.y - bp.y) * (dp.y - bp.y) < 36
        RETURN bp.frame AS frame, count(d) AS pressers
        ORDER BY bp.frame
    """))
    observed_frames = {int(r["frame"]): int(r["pressers"])
                       for r in presser_rows
                       if int(r["pressers"]) >= 3}
    print(f"  {len(observed_frames)} observed frames met threshold")

    # B.2: predicted coach-intent
    print("\n[predicted] Coach-intent pressing plan(s):")
    for row in db.execute("""
        MATCH (c:CoachIntent)
        RETURN c.intent_id AS id, c.trigger AS trigger, c.action AS action,
               c._confidence AS conf, c._source AS source
    """):
        print(f"  {dict(row)}")

    # B.3: predicted LLM commentary tags around pressing frames
    print("\n[predicted] LLM commentary tags within ±10 frames of a pressing-trigger frame:")
    if observed_frames:
        sample_frame = sorted(observed_frames.keys())[0]
        for row in db.execute(f"""
            MATCH (c:CommentaryTag)
            WHERE c.frame >= {sample_frame - 10} AND c.frame <= {sample_frame + 10}
            RETURN c.frame AS frame, c.tag AS tag,
                   c._confidence AS conf, c._source AS source
            ORDER BY c.frame
        """):
            print(f"  {dict(row)}")

    print("\n--- What this proves ---")
    print("  - Tactical reconstruction is one MATCH per event AS OF its frame —")
    print("    no separate audit table, no video re-processing job.")
    print("  - The three confidence tiers coexist in one graph; the same query")
    print("    shape spans observed tracking, predicted coach-intent, and")
    print("    predicted LLM commentary. Trust-tier filtering is one predicate.")
    print("  - When the next signal source arrives — referee-decision tracker,")
    print("    biomechanical fatigue model, opponent-scouting database — it")
    print("    writes _observation_class facts into the same graph and every")
    print("    existing fusion query absorbs it without schema migration.")

    db.close()


if __name__ == "__main__":
    main()

File: examples/team-sports-tactical-world-model/rust/src/bin/03-tactical-replay-and-fusion.rs
//! Step 3 (Rust) — Tactical AS OF replay, fusion, and live-alert.
//!
//! Mirrors `03-tactical-replay-and-fusion.py` for the replay + fusion
//! parts, AND reaches the Rust-only surface for the live-alert pattern
//! Python can't do today: `arcflow_sdk::register_skill` fires the named
//! tactical-pattern query every time a position mutation matches the
//! trigger, with no polling loop.

use anyhow::Result;
use arcflow_sdk::SkillTrigger;
use team_sports_tactical_world_model::{
    make_db, EVENT_COUNTER_ORIGIN, EVENT_LINE_BREAK, EVENT_PRESS_START,
};

fn main() -> Result<()> {
    let db = make_db()?;

    // ---- A. Tactical AS OF reconstruction -----------------------------
    println!("=== Tactical AS OF reconstruction ===");
    for (label, frame) in [
        ("pressing trigger",       EVENT_PRESS_START + 6),
        ("line-breaking pass",     EVENT_LINE_BREAK),
        ("counter-attack origin",  EVENT_COUNTER_ORIGIN + 8),
    ] {
        println!("\n--- {label} @ frame {frame} ---");
        let q = format!(
            "MATCH (b:Ball)-[:POSITION_AT]->(p:Position) \
             WHERE p.frame = {frame} \
             RETURN p.x AS x, p.y AS y, p.team_possession AS poss"
        );
        for row in db.execute(&q)?.rows {
            println!("  Ball: {row:?}");
        }
    }

    // ---- B. Register a reactive skill for live alert (Rust-only) ------
    // This is the part Python can't reach today. The skill fires every
    // time a Position node is created that matches the trigger; the named
    // query computes the pressing count and surfaces the result without
    // a polling loop.
    println!("\n=== Live tactical alert via register_skill (Rust-only) ===");
    let trigger = SkillTrigger::NodeCreated {
        label: "Position".to_string(),
    };
    let alert_q = "MATCH (b:Ball)-[:POSITION_AT]->(bp:Position), \
                         (d:Player {team: 'defending'})-[:POSITION_AT]->(dp:Position) \
                   WHERE dp.frame = bp.frame \
                     AND (dp.x - bp.x) * (dp.x - bp.x) + (dp.y - bp.y) * (dp.y - bp.y) < 36 \
                   WITH bp.frame AS frame, count(d) AS pressers \
                   WHERE pressers >= 3 \
                   RETURN frame, pressers";
    arcflow_sdk::register_skill(
        &db,
        "tactics",
        "pressing-alert",
        trigger,
        alert_q,
    )?;
    println!("  registered: tactics/pressing-alert");
    println!("  trigger:    NodeCreated label='Position'");
    println!("  fires:      when >=3 defenders within 6m of ball");
    println!("  surface:    every matching mutation surfaces the alert without polling");

    println!("\n--- What this proves ---");
    println!("  - Frame-resolution reconstruction is one MATCH per event.");
    println!("  - The live-alert pattern (register_skill) is the Rust-only surface");
    println!("    that closes the loop from 'detect the pattern' to 'fire on the");
    println!("    mutation that triggered it' — no polling, no cron, no scheduler.");

    Ok(())
}
